ls1intum/Artemis
# `Tutorial groups`: Allow instructors to create tutorial groups

This feature is currently hidden based on a feature toggle




### Checklist
#### General


#### Server
#### Client

### Motivation and Context
- First important step for the management of tutorial groups in Artemis : Basic entity and related CRUD(Create,Read,Update,Delete) functionality
- **Functionality included in this PR**:

1. Create,Update,Delete Tutorial Groups with basic properties like title, teaching assistant and so on
2. Registered Students management + import / export csv

**Remaining functionality like Session Management (When does a tutorial group happen?) will be done in further PRs**

### Description
- Removed existing TutorGroup entity and database tables
- Added new TutorialGroup entity and database tables
- Implemented CRUD functionality for TutorialGroups
- **Made the course group component reusable for the management of registered students of tutorial groups**

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Feature Toggle for TutorialGroups activated

**Testing is pretty simple. Just create some tutorial groups, edit the information, delete them and register/deregister some students**


**Important: Because I reused the component for adding students, tutors and so on to courses for the management of registered students for a tutorial group it would be wise to test if adding students, tutors and so on to courses also still works as expected!**

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| TutorialGroupResource.java | 100% | 100% |
| TutorialGroupService.java | 100% | 100% |
| tutorial-groups.service.ts | 100% | 100% |
| tutorial-groups-management.component.ts | 100% | 100% |
| tutorial-groups-management.component.ts | 100% | 100% |
| registered-students.component.ts | 73% | 93% |
| tutorial-group-form.component.ts | 69% | 91% |
| edit-tutorial-group.component.ts | 75% | 95% |
| create-tutorial-group.component.ts | 73% | 93% |
| tutorial-group-row-buttons.component.ts| 100% | 100% |


### Screenshots
![image](https://user-images.githubusercontent.com/29718932/185785808-73c75c2f-b136-442e-b743-7e6ff9cf3930.png)
![image](https://user-images.githubusercontent.com/29718932/185785841-13d7b51e-e571-4f52-85d9-a3271b445f49.png)
![image](https://user-images.githubusercontent.com/29718932/185785855-a3de1052-3b36-4e18-a20e-04efbbd705c7.png)
![image](https://user-images.githubusercontent.com/29718932/185785870-972cecab-9aa3-4421-adbd-1da58e3a12c4.png)
![image](https://user-images.githubusercontent.com/29718932/185785876-ca873be1-4cf5-404f-baae-af1272a95189.png)


### DB Changes
Note: The registration entity has been introduced because we want to allow self-registration of students later. 

![crud_model](https://user-images.githubusercontent.com/29718932/190107717-63f7f907-ac23-47dd-b563-16d7541fe1aa.png)
